---
title: 'spatialDEG: installation and simulation'
author: "Yi Yang"
date: "`r Sys.Date()`"
output:
  html_document: default
  word_document: default
  pdf_document: default
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Install spatialDEG
This vignette provides an introduction to the R package `spatialDEG`, where the function `spatialDEG` implements the model `spatialDEG`,  Identification of differentially expressed genes by leveraging spatial information in spatially resolved transcriptomic studies.
The package can be installed with the command:

`library(devtools)`

`install_github("Shufeyangyi2015310117/spatialDEG")`

The package can be loaded with the command:
```{r}
library("spatialDEG")
```

## Perform spatialDEG analysis on simulated data
### Generating the simulated data
We first set the basic parameter:
```{r echo=TRUE, message=FALSE, warning=FALSE}
set.seed(100)
n = 400
ngene <- 5
he = 0.1
num_ls = 80
## spaital location
spa <- cbind(rep(1:20, 20), rep(1:20, each=20))
```

Then, we generate the Gaussian kernel. 
```{r}
## generate kernel, which controls spatial component
sq_dist <- get_sq_dist(spa)
scale_q <- 0.3
scale_param <- sqrt( quantile(sq_dist[sq_dist > 0.0], scale_q) ) # larger
Kernel <- get_K_gauss(sq_dist, scale_param)
res <- SysMatEigen2(Kernel)
K <- res$kernel_mat
e <- eigen(K)
V <- e$vectors
B <- V %*% diag(sqrt(e$values))
```                    

Then, we generate spatial component and non-spatial component.
```{r}
library(MASS)
# non-spatial component
se <- matrix(0, nrow=ngene, ncol=ngene)
diag(se) <- 1
vecE <- mvrnorm(n, mu=rep(0,ngene), Sigma=se)

# spatial component
vecB <- t(mvrnorm(ngene, mu=rep(0,n), Sigma=K))

vecY = matrix(0, n, ngene)
# total = spatial + non-spatial component
for (j in 1:ngene){
    vecY[,j] <- vecB[,j]*sqrt(var(vecE[,j])/var(vecB[,j]))*sqrt(he/(1-he)) + vecE[,j]
}
Y = vecY
                    
                    
# non-spatial component
se <- matrix(0, nrow=ngene, ncol=ngene) # covariance between genes = seab, var = sea
diag(se) <- 1
vecE <- mvrnorm(n, mu=rep(0,ngene), Sigma=se)

# spatial component
vecB <- t(mvrnorm(ngene, mu=rep(0,n), Sigma=K))
```

Then, we generate the gene expression.
```{r}
vecY = matrix(0, n, ngene)
# total = spatial + non-spatial component
for (j in 1:ngene){
    vecY[,j] <-vecB[,j]*sqrt(var(vecE[,j])/var(vecB[,j]))*sqrt(he/(1-he)) + vecE[,j]
}
b = 5

K2 = K
Y1 = vecY + b
Y2 = vecY + b
W1 = matrix(1, n, 1)
W2 = matrix(1, n, 1)
```


### Fitting spatialDEG
Here, we set the basic arguments of `spatialDEG`
```{r}
max_Iter = 100
```
Here, we briefly explain the arguments of `spatialDEG`.

'Y1' is a vector of gene expression for the first dataset.

'Y2' is a vector of gene expression for the second dataset.

'spa1' is a matrix of spatial location for the first dataset. If a kernel is provided for the argument 'Kernel1', spatialDEG automatically ignores this argument.

'spa2' is a matrix of spatial location for the second dataset. If a kernel is provided for the argument 'Kernel2', spatialDEG automatically ignores this argument.

'W1' is a matrix of covariates for the first dataset.

'W2' is a matrix of covariates for the second dataset.

'Initial_theta' is a vector of initial value for theta. The default vector is a vector of ones. 

'num_ls' is a integer specifying the total number of length scale and periodicity parameters. The default is 10. If kernels are provided for the arguments 'Kernel1' and 'Kernel2', spatialDEG automatically ignores this argument.

'Kernel1' is a matrix of Kernel for the first dataset.

'Kernel2' is a matrix of Kernel for the second dataset.

'parallel' is a logical value specifying whether to run spatialDEG in parallel. The default is True.

'coreNum' is an integer specifying the number of cores used for parallel computation. The default is 1.

'max_Iter' is an integer specifying the maximum iteration of the AI algorithm. The default is 1000.

'eps' is a value specifying when the AI algorithm convergence.
'kernel_fixed' Kernel_fixd a logical value specifying whether the kernel is fixed

'kernel_matched' a logical value specifying whether the the kernels under both null hypothesis and alternative hypothesis are restricted to be the same type of kernel. The default is FALSE.

'check_positive' a logical value specifying whether to check the positive definiteness of kernels. The default is TRUE.


#### Run spatialDEG with the estimated kernels
Finally, we run `spatialDEG` with the estimated kernels
```{r message=FALSE, warning=FALSE}
parallel = FALSE
coreNum = 2
fit = spatialDEG(Y1, Y2, spa, spa, W1, W2, coreNum = coreNum, parallel = parallel, max_Iter = max_Iter)
print(fit)
```
Here, We briefly explain the output of `spatialDEG`. 

The item 'chi_stat' is chi-square test statistics of the log-likelihood ratio test.

The item 'pvalue' is p value of the log-likelihood ratio test.

The item 'sigma_e_alternative' is the estimated variance corresponding to identity error under the alternative hypothesis.

The item 'sigma_kernel1_alternative' is the estimated variance corresponding to the first kernel under the alternative hypothesis.

The item 'sigma_kernel2_alternative' is the estimated variance corresponding to the second kernel under the alternative hypothesis.

The item 'beta_alternative' is the estimated coefficients corresponding to covariates under the alternative hypothesis.

The item 'mu' is the coefficients corresponding to indicators.

The item 'sigma_e_null' is the estimated variance corresponding to identity error under the null hypothesis.

The item 'sigma_kernel1_null' is the estimated variance corresponding to the first kernel under the null hypothesis.

The item 'sigma_kernel2_null' is the estimated variance corresponding to the second kernel under the null hypothesis.

The item 'beta_null' is the estimated coefficients corresponding to covariates under the null hypothesis.

The item 'kernel1_all_vec' is a predefined length scale or periodicity for the first dataset.

The item 'kernel2_all_vec' is a predefined length scale or periodicity for the second dataset.

The item 'kernel1_length_scale_periodicity_alternative' is the estimated length scale of the Gaussian kernel or periodicity of the Cosine kernel for the first dataset under the alternative hypothesis.

The item 'kernel2_length_scale_periodicity_alternative' is the estimated length scale of the Gaussian kernel or periodicity of the Cosine kernel for the second dataset under the alternative hypothesis.

The item 'kernel1_length_scale_periodicity_null' is the estimated length scale of the Gaussian kernel or periodicity of the Cosine kernel for the first dataset under the null hypothesis.

The item 'kernel2_length_scale_periodicity_null' is the estimated length scale of the Gaussian kernel or periodicity of the Cosine kernel for the second dataset under the null hypothesis.

The item 'kernel1_length_scale_periodicity_idx_null' is the index of the estimated length scale for the Gaussian kernel or the periodicity for the Cosine kernel in the vector of kernel1_all_vec

The item 'kernel2_length_scale_periodicity_idx_null' is the index of the estimated length scale for the Gaussian kernel or the periodicity for the Cosine kernel in the vector of kernel2_all_vec


#### Run spatialDEG with the estimated kernels in parallel
We also can run `spatialDEG` with the estimated kernels in parallel
```{r message=FALSE, warning=FALSE}
parallel = TRUE
fit = spatialDEG(Y1, Y2, spa, spa, W1, W2, parallel = parallel, max_Iter = max_Iter)
print(fit)
```
The items of output are defined the same as the items above.

#### Select the genes that have the same type of kernel under both null and alternative hypothesis. 
```{r message=FALSE, warning=FALSE}
result = select_gene(fit)
print(result)
```
The items of output are defined the same as the items above.

#### Run spatialDEG with the kernels specified by the user
When the true kernel is available, we can run `spatialDEG` with the kernels specified by the user.
```{r message=FALSE, warning=FALSE}
parallel = FALSE
fit = spatialDEG(Y1 = Y1, Y2 = Y2, W1 = W1, W2 = W2, Kernel1 = K, Kernel2 = K2, parallel = parallel, max_Iter = max_Iter)
print(fit)
```
The items of output are defined the same as the items above.