# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @author Yi Yang
#' @title
#' SysMatEigen2
#' @description
#' SysMatEigen2 to transform the matrix to a positive definite matrix
#' @param M  a matrix
#' @return a List
#'
#'
#' @details
#' \code{spatialDEG_test} run spatialDEG model.
#' @export
SysMatEigen2 <- function(M) {
    .Call(`_spatialDEG_SysMatEigen2`, M)
}

#' @author Yi Yang
#' @title
#' spatialDEG
#' @description
#' spatialDEG to perform differentially expressed gene analysis by leveraging spatial information with true kernels.
#' @param W  covariates for the two datasets
#' @param Y  stacked gene expression
#' @param K1  kernel for the first dataset.
#' @param K2  kernel for the second dataset.
#' @param Initial_theta  initial value of theta
#' @param mu_fixed logical value specifying whether mu is fixed
#' @param verbose logical value specifying whether to print the parameters during the spatialDEG analysis or not
#' @param check_positive a logical value specifying whether to check the positive definiteness of kernels
#'
#' @return List of model parameters
#'
#'
#' @details
#' \code{spatialDEG_test} run spatialDEG model.
#' @export
spatialDEG_true_kernel_test <- function(W, Y, K1, K2, Initial_theta, mu_fixed = TRUE, verbose = TRUE, check_positive = FALSE) {
    .Call(`_spatialDEG_spatialDEG_true_kernel_test`, W, Y, K1, K2, Initial_theta, mu_fixed, verbose, check_positive)
}

#' @author Yi Yang
#' @title
#' spatialDEG_paral_test
#' @description
#' spatialDEG to perform differentially expressed gene analysis by leveraging spatial information in parallel.
#'
#' @param spa1  spatial location for the first dataset.
#' @param spa2  spatial location for the second dataset.
#' @param W  covariates for the two datasets
#' @param Y  stacked gene expression
#' @param Initial_theta  initial value of theta
#' @param num_ls total number of length scale and periodicity
#' @param iteration The maximum iteration
#' @param Kernel_fixd a logical value specifying whether the kernel is fixed
#' @param verbose a logical value specifying whether to print the parameters during the spatialDEG analysis or not
#' @param kernel_matched a logical value specifying whether the the kernels under both null hypothesis and alternative hypothesis are restricted to be the same type of kernel
#' @param coreNum Number of cores used for parallel computation
#' @param tol tolence to stop the algorithm
#'
#' @return List of model parameters
#'
#'
#' @details
#' \code{spatialDEG_paral_test} run spatialDEG model.
#' @export
spatialDEG_paral_test <- function(spa1, spa2, W, Y, Initial_theta, num_ls = 20L, iteration = 100L, Kernel_fixd = FALSE, verbose = FALSE, kernel_matched = FALSE, coreNum = 1L, tol = 0.0001) {
    .Call(`_spatialDEG_spatialDEG_paral_test`, spa1, spa2, W, Y, Initial_theta, num_ls, iteration, Kernel_fixd, verbose, kernel_matched, coreNum, tol)
}

#' @author Yi Yang
#' @title
#' spatialDEG
#' @description
#' spatialDEG to perform differentially expressed gene analysis by leveraging spatial information.
#'
#' @param spa1  spatial location for the first dataset.
#' @param spa2  spatial location for the second dataset.
#' @param W  covariates for the two datasets
#' @param Y  stacked gene expression
#' @param Initial_theta  initial value of theta
#' @param num_ls total number of length scale and periodicity
#' @param mu_fixed logical value specifying whether mu is fixed
#' @param verbose logical value specifying whether to print the parameters during the spatialDEG analysis or not
#'
#' @return List of model parameters
#'
#'
#' @details
#' \code{spatialDEG_test} run spatialDEG model.
#' @export
spatialDEG_test <- function(spa1, spa2, W, Y, Initial_theta, num_ls = 20L, mu_fixed = TRUE, verbose = TRUE) {
    .Call(`_spatialDEG_spatialDEG_test`, spa1, spa2, W, Y, Initial_theta, num_ls, mu_fixed, verbose)
}

